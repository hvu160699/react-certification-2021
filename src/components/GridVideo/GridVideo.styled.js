import styled from '@emotion/styled';

const GridVideoContainer = styled.div`
  max-width: 100%;
  display: grid;
  ${(props) =>
    props.vertical
      ? `grid-template-columns: repeat(1, minmax(0, 1fr));
         gap: 1rem 0;
        `
      : `grid-template-columns: repeat(1, minmax(0, 1fr));
        gap: 1rem;
      `}

  @media (min-width: 768px) {
    grid-template-columns: ${(props) =>
      props.vertical ? 'auto' : 'repeat(2, minmax(0, 1fr))'};
  }

  @media (min-width: 1024px) {
    grid-template-columns: ${(props) =>
      props.vertical ? 'auto' : 'repeat(3, minmax(0, 1fr))'};
  }

  @media (min-width: 1440px) {
    grid-template-columns: ${(props) =>
      props.vertical ? 'auto' : 'repeat(4, minmax(0, 1fr))'};
  }
`;

const GridVideoItem = styled.div`
  display: flex;
  flex-direction: ${(props) => (props.vertical ? 'row' : 'column')};
  justify-content: flex-start;
  align-items: stretch;
  width: 100%;
  height: 100%;
  overflow: hidden;

  .thumbnail {
    display: block;
    position: relative;
    flex: none;
    ${(props) =>
      props.vertical
        ? `width: 168px;
    margin-right: 0.5rem;`
        : `width: 100%;
          margin-bottom: 0.5rem;
        `}
    height: 94px;

    @media (min-width: 640px) {
      height: ${(props) => (props.vertical ? '' : '188px')};
    }

    .img-wrapper {
      display: block;
      height: 100%;
      overflow: hidden;
      margin-left: auto;
      margin-right: auto;
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      background: #000000;

      img {
        display: block;
        width: 100%;
        max-height: none;
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
        position: absolute;
        transform: translateY(-50%);
        top: 50%;
        left: 0;
      }
    }
  }

  .content {
    width: 100%;
    min-width: 0;
    display: flex;
    flex-direction: column;

    h3 {
      margin: 0 0 4px 0;

      span {
        display: -webkit-box;
        -webkit-line-clamp: 2;
        -webkit-box-orient: vertical;
        max-height: inherit;
        overflow: hidden;
        color: #111827;
        text-overflow: ellipsis;
        white-space: normal;
        font-size: 0.875rem;
        font-weight: 500;
        line-height: 1.2rem;
        letter-spacing: 0.25px;
      }
    }

    .statistics {
      display: flex;
      flex-direction: column;
      flex-wrap: wrap;

      .row {
        max-width: 100%;
        display: flex;
        flex-direction: row;
        align-items: center;
        flex-wrap: wrap;
        overflow: hidden;

        span {
          color: gray;
          font-size: 0.75rem;
          line-height: 1rem;
          display: inline-block;
          align-self: flex-start;
          text-overflow: ellipsis;
          white-space: hidden;
          overflow: hidden;
        }
      }
    }
  }
`;

GridVideoContainer.displayName = 'GridVideoContainer';
GridVideoItem.displayName = 'GridVideoItem';

export { GridVideoContainer, GridVideoItem };
