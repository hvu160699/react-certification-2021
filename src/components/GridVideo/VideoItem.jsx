import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';

const VideoItemContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: stretch;
  height: 100%;
  overflow: hidden;
  border-radius: 0.25rem;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06);
  cursor: pointer;
  transition: all 250ms linear;

  img {
    object-fit: cover;
    width: 100%;
    height: 15rem;
    max-width: 100%;
  }

  .content {
    padding: 0 1rem;

    h1 {
      font-size: 1.25em;
      font-weight: 500;
      color: #111827;
    }

    p {
      margin-top: 0;
      font-weight: 400;
      color: #6b7280;
    }
  }

  &:hover {
    box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05);
  }
`;

const VideoItem = ({ video }) => {
  const { title, description, thumbnails } = video;
  return (
    <VideoItemContainer>
      <picture>
        <source srcSet={thumbnails.high.url} media="(min-width: 1023.98px)" />
        <img loading="lazy" src={thumbnails.medium.url} alt={title} />
      </picture>
      <div className="content">
        <h1>{title || 'No title'}</h1>
        <p>{description || ''}</p>
      </div>
    </VideoItemContainer>
  );
};

VideoItem.propTypes = {
  video: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default memo(VideoItem);
