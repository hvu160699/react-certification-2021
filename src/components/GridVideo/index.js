export { default } from './GridVideo.component';
