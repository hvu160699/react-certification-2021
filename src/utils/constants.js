const AUTH_STORAGE_KEY = 'wa_cert_authenticated';
const API_URL = process.env.REACT_APP_API_URL;
const DEFAULT_AVATAR =
  'https://images.unsplash.com/photo-1491528323818-fdd1faba62cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80';

const FAKE_VIDEOS_DATA = [
  {
    kind: 'youtube#searchResult',
    etag: '_PVKwNJf_qw9nukFeRFOtQ837o0',
    id: {
      kind: 'youtube#channel',
      channelId: 'UCPGzT4wecuWM0BH9mPiulXg',
    },
    snippet: {
      publishedAt: '2014-09-27T01:39:18Z',
      channelId: 'UCPGzT4wecuWM0BH9mPiulXg',
      title: 'Wizeline',
      description:
        "Wizeline transforms how teams build technology. Its customers accelerate the delivery of innovative products with proven solutions, which combine Wizeline's ...",
      thumbnails: {
        default: {
          url:
            'https://yt3.ggpht.com/ytc/AAUvwnighSReQlmHl_S_vSfvnWBAG5Cw4A0YxtE0tm5OpQ=s88-c-k-c0xffffffff-no-rj-mo',
        },
        medium: {
          url:
            'https://yt3.ggpht.com/ytc/AAUvwnighSReQlmHl_S_vSfvnWBAG5Cw4A0YxtE0tm5OpQ=s240-c-k-c0xffffffff-no-rj-mo',
        },
        high: {
          url:
            'https://yt3.ggpht.com/ytc/AAUvwnighSReQlmHl_S_vSfvnWBAG5Cw4A0YxtE0tm5OpQ=s800-c-k-c0xffffffff-no-rj-mo',
        },
      },
      channelTitle: 'Wizeline',
      liveBroadcastContent: 'upcoming',
      publishTime: '2014-09-27T01:39:18Z',
    },
  },
  {
    kind: 'youtube#searchResult',
    etag: 'erqeM78PZDWIBe8qOGHGM2WdSE8',
    id: {
      kind: 'youtube#video',
      videoId: 'nmXMgqjQzls',
    },
    snippet: {
      publishedAt: '2019-09-30T23:54:32Z',
      channelId: 'UCPGzT4wecuWM0BH9mPiulXg',
      title: 'Video Tour | Welcome to Wizeline Guadalajara',
      description:
        'Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...',
      thumbnails: {
        default: {
          url: 'https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg',
          width: 120,
          height: 90,
        },
        medium: {
          url: 'https://i.ytimg.com/vi/nmXMgqjQzls/mqdefault.jpg',
          width: 320,
          height: 180,
        },
        high: {
          url: 'https://i.ytimg.com/vi/nmXMgqjQzls/hqdefault.jpg',
          width: 480,
          height: 360,
        },
      },
      channelTitle: 'Wizeline',
      liveBroadcastContent: 'none',
      publishTime: '2019-09-30T23:54:32Z',
    },
  },
];

export { AUTH_STORAGE_KEY, API_URL, DEFAULT_AVATAR, FAKE_VIDEOS_DATA };
