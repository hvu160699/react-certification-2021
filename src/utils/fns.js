import { lazy } from 'react';

function random(limit) {
  return Math.floor(Math.random() * limit);
}

function loader(pageName) {
  return lazy(() => import(`../pages/${pageName}`));
}

function createQuerystring(queryData) {
  const queryStringArray = Object.keys(queryData).reduce((acc, key) => {
    let value = '';

    if (Array.isArray(queryData[key])) value = queryData[key].join(',');
    else value = queryData[key];

    acc.push(`${key}=${value}`);
    return acc;
  }, []);

  let queryString = '';
  if (queryStringArray.length) {
    queryString = `?${queryStringArray.join('&')}`;
  }

  return queryString;
}

export { random, loader, createQuerystring };
