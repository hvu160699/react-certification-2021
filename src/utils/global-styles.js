import React from 'react';
import { Global, css } from '@emotion/react';

const GlobalStyles = () => {
  return (
    <Global
      styles={css`
        html,
        body,
        #root {
          height: 100%;
        }

        body {
          font-family: 'Ubuntu', sans-serif;
        }

        *,
        *:after,
        *:before {
          -webkit-box-sizing: border-box;
          -moz-box-sizing: border-box;
          box-sizing: border-box;
        }

        .sr-only {
          position: absolute;
          width: 1px;
          height: 1px;
          padding: 0;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          border: 0;
        }

        svg {
          width: 1.5rem;
          height: 1.5rem;
        }
      `}
    />
  );
};

export default GlobalStyles;
