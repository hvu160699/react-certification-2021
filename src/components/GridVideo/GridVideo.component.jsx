import React from 'react';
import PropTypes from 'prop-types';

import VideoItem from './GridVideoItem.component';

import { GridVideoContainer } from './GridVideo.styled';

const GridVideo = ({ videos }) => {
  return (
    <GridVideoContainer>
      {videos && videos.length > 0 ? (
        videos.map((item) => <VideoItem key={item.etag} video={item.snippet} />)
      ) : (
        <div>No result</div>
      )}
    </GridVideoContainer>
  );
};

GridVideo.propTypes = {
  videos: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default GridVideo;
