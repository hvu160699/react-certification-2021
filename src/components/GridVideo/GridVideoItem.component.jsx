import React, { memo } from 'react';
import PropTypes from 'prop-types';

import { GridVideoItem } from './GridVideo.styled';

const VideoItem = ({ video }) => {
  const { title, description, thumbnails } = video;
  return (
    <GridVideoItem>
      <picture>
        <source srcSet={thumbnails.high.url} media="(min-width: 1023.98px)" />
        <img loading="lazy" src={thumbnails.medium.url} alt={title} />
      </picture>
      <div className="content">
        <h1>{title}</h1>
        <p>{description}</p>
      </div>
    </GridVideoItem>
  );
};

VideoItem.propTypes = {
  video: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default memo(VideoItem);
