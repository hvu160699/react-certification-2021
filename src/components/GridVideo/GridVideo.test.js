import React from 'react';

import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { FAKE_VIDEOS_DATA } from '../../utils/constants';

import GridVideo from './GridVideo.component';
import GridVideoItem from './GridVideoItem.component';

describe('Grid Video, Grid Video Item', () => {
  it('<GridVideo /> shoud render "No result" when passing an empty array', () => {
    const gridVideo = shallow(<GridVideo videos={[]} />);
    expect(gridVideo.text()).toEqual('No result');
  });

  it('<GridVideo /> should render as much <GridVideoItem /> as FAKE_VIDEOS_DATA', () => {
    const gridVideo = shallow(<GridVideo videos={FAKE_VIDEOS_DATA} />);
    expect(gridVideo.find(GridVideoItem).length).toBe(FAKE_VIDEOS_DATA.length);
  });

  it('<GridVideoItem /> should render without crashing', () => {
    const gridVideoItem = renderer
      .create(<GridVideoItem video={FAKE_VIDEOS_DATA[0].snippet} />)
      .toJSON();
    expect(gridVideoItem).toMatchSnapshot();
  });
});
