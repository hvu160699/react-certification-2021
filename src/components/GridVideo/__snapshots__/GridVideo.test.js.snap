// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Video, Grid Video Item <GridVideoItem /> should render without crashing 1`] = `
<div
  className="css-istpc4"
>
  <picture>
    <source
      media="(min-width: 1023.98px)"
      srcSet="https://yt3.ggpht.com/ytc/AAUvwnighSReQlmHl_S_vSfvnWBAG5Cw4A0YxtE0tm5OpQ=s800-c-k-c0xffffffff-no-rj-mo"
    />
    <img
      alt="Wizeline"
      loading="lazy"
      src="https://yt3.ggpht.com/ytc/AAUvwnighSReQlmHl_S_vSfvnWBAG5Cw4A0YxtE0tm5OpQ=s240-c-k-c0xffffffff-no-rj-mo"
    />
  </picture>
  <div
    className="content"
  >
    <h1>
      Wizeline
    </h1>
    <p>
      Wizeline transforms how teams build technology. Its customers accelerate the delivery of innovative products with proven solutions, which combine Wizeline's ...
    </p>
  </div>
</div>
`;
